### ISSUE TITLE
-Wconditional-uninitialized false positive

### ISSUE THREAD

**Issue by timblechmann** [llvm/llvm-project#91438]
broken out of qt's `QColorTransferFunction`:

```
struct foo {

    void bar()
    {
        float a;
        if (m_b)
            a = std::pow(1.0f / m_a, 1.f);
        if (m_b)
            std::isfinite(a);
        if (m_b && std::isfinite(a))
            return;
    }

    float m_a;
    bool m_b;
};
```

`-Wconditional-uninitialized` warns about `a` not being initialized in `std::isfinite(a)`, but in both cases the variable is initialized

**Comment by timblechmann** [comment]
https://godbolt.org/z/achfh9hTn

**Comment by MitalAshok** [comment]
Related examples that warn right now but maybe shouldn't:

```c++
void foo(const bool cond) {
    int a;
    if (cond)
        a = 243;

    if (cond)
        (void)+a;
}

void bar(const bool cond) {
    int a;
    if (cond)
        a = 243;
    if (!cond)
        a = 6;

    (void)+a;
}
```

Your example is a bit trickier since the object parameter is not local to the function, thus `m_b` could be modified somewhere else to mean you do access `a` uninitialized. For this specific case, we know that it is unchanged because no functions are called between the two checks (if `m_b` is `false`), but that's not true in the general case so it might not be worth implementing.

**Comment by shafik** [comment]
Even if we set `m_b` it still does not get this right.

**Comment by shafik** [comment]
Looks like a duplicate of: https://github.com/llvm/llvm-project/issues/38204

Which explains this is expected short-coming of this diagnostic.

**Comment by timblechmann** [comment]
thanks @shafik for looking into it.
